### ISSUE TITLE
Different results from FindMarkers by Seurat Version 4.2.0 than 4.1.0

### ISSUE THREAD

**Issue by Yuwei0128** [satijalab/seurat#6586]
Dear Seurat team,

I updated the Seurat package to version 4.2.0 and noticed that the results by FindMarkers and FindAllMarkers were different than ones generated by Seurat v4.1.0. 

I loaded the old Rds file generated by Seurat v4.1.0 and re-calculated the differentially expressed genes by FindMarkers from Seurat 4.2.0. Less markers were identified with smaller avg_log2FC values. 

The workflow was same as the introduction to SCTransform v2 regularization. 

After the deletion and re-installation of Seurat v4.1.0, the result was same as before. And the session info (Seurat v4.1) was 

R version 4.1.2 (2021-11-01)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Monterey 12.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] enrichR_3.1                 MAST_1.20.0                 DoMultiBarHeatmap_0.1.0     magrittr_2.0.3             
 [5] rlang_1.0.6                 SingleCellExperiment_1.16.0 SummarizedExperiment_1.24.0 Biobase_2.54.0             
 [9] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1         IRanges_2.28.0              S4Vectors_0.32.4           
[13] BiocGenerics_0.40.0         MatrixGenerics_1.6.0        matrixStats_0.62.0          dittoSeq_1.6.0             
[17] ggplot2_3.3.6               Scillus_0.5.0               DT_0.25                     patchwork_1.1.2            
[21] dplyr_1.0.10                stringr_1.4.1               SeuratData_0.2.2            Seurat_4.1.0               
[25] sp_1.5-0                    SeuratObject_4.1.2         

loaded via a namespace (and not attached):
  [1] utf8_1.2.2             reticulate_1.26        tidyselect_1.2.0       htmlwidgets_1.5.4      grid_4.1.2            
  [6] Rtsne_0.16             munsell_0.5.0          codetools_0.2-18       ica_1.0-3              future_1.28.0         
 [11] miniUI_0.1.1.1         withr_2.5.0            spatstat.random_2.2-0  colorspace_2.0-3       progressr_0.11.0      
 [16] knitr_1.40             rstudioapi_0.14        ROCR_1.0-11            tensor_1.5             listenv_0.8.0         
 [21] GenomeInfoDbData_1.2.7 polyclip_1.10-0        pheatmap_1.0.12        parallelly_1.32.1      vctrs_0.4.2           
 [26] generics_0.1.3         xfun_0.33              R6_2.5.1               doParallel_1.0.17      clue_0.3-61           
 [31] bitops_1.0-7           spatstat.utils_2.3-1   cachem_1.0.6           DelayedArray_0.20.0    assertthat_0.2.1      
 [36] promises_1.2.0.1       scales_1.2.1           rgeos_0.5-9            gtable_0.3.1           globals_0.16.1        
 [41] processx_3.7.0         goftest_1.2-3          GlobalOptions_0.1.2    splines_4.1.2          lazyeval_0.2.2        
 [46] spatstat.geom_2.4-0    formattable_0.2.1      yaml_2.3.6             reshape2_1.4.4         abind_1.4-5           
 [51] crosstalk_1.2.0        httpuv_1.6.6           tools_4.1.2            ellipsis_0.3.2         spatstat.core_2.4-4   
 [56] jquerylib_0.1.4        RColorBrewer_1.1-3     ggridges_0.5.4         Rcpp_1.0.9             plyr_1.8.7            
 [61] zlibbioc_1.40.0        purrr_0.3.5            RCurl_1.98-1.9         ps_1.7.1               prettyunits_1.1.1     
 [66] rpart_4.1.16           deldir_1.0-6           pbapply_1.5-0          GetoptLong_1.0.5       cowplot_1.1.1         
 [71] zoo_1.8-11             ggrepel_0.9.1          cluster_2.1.4          data.table_1.14.4      scattermore_0.8       
 [76] circlize_0.4.15        lmtest_0.9-40          RANN_2.6.1             fitdistrplus_1.1-8     mime_0.12             
 [81] evaluate_0.17          xtable_1.8-4           gridExtra_2.3          shape_1.4.6            compiler_4.1.2        
 [86] tibble_3.1.8           KernSmooth_2.23-20     crayon_1.5.2           htmltools_0.5.3        mgcv_1.8-40           
 [91] later_1.3.0            tidyr_1.2.1            DBI_1.1.3              ComplexHeatmap_2.13.2  MASS_7.3-58.1         
 [96] rappdirs_0.3.3         Matrix_1.5-1           cli_3.4.1              parallel_4.1.2         igraph_1.3.5          
[101] forcats_0.5.2          pkgconfig_2.0.3        plotly_4.10.0          spatstat.sparse_2.1-1  foreach_1.5.2         
[106] bslib_0.4.0            XVector_0.34.0         callr_3.7.2            digest_0.6.30          sctransform_0.3.5     
[111] RcppAnnoy_0.0.19       spatstat.data_2.2-0    rmarkdown_2.17         leiden_0.4.3           uwot_0.1.14           
[116] curl_4.3.3             shiny_1.7.2            rjson_0.2.21           lifecycle_1.0.3        nlme_3.1-160          
[121] jsonlite_1.8.2         viridisLite_0.4.1      limma_3.50.3           fansi_1.0.3            pillar_1.8.1          
[126] lattice_0.20-45        fastmap_1.1.0          httr_1.4.4             pkgbuild_1.3.1         survival_3.4-0        
[131] glue_1.6.2             remotes_2.4.2          png_0.1-7              iterators_1.0.14       stringi_1.7.8         
[136] sass_0.4.2             irlba_2.3.5.1          future.apply_1.9.1  

Thank you.

**Comment by timoast** [comment]
Can you show the code you're using for differential expression testing?

**Comment by Yuwei0128** [comment]
#Integration
ob.list <- PrepSCTIntegration(object.list = ob.list, anchor.features = Newfeature_anchor)
immune.anchors <- FindIntegrationAnchors(object.list = ob.list, normalization.method = "SCT",
                                         anchor.features = Newfeature_anchor)
immune.combined.sct <- IntegrateData(anchorset = immune.anchors, normalization.method = "SCT")

#Dimension reductio
immune.combined.sct <- RunPCA(immune.combined.sct, npcs = 40,verbose = FALSE)
ElbowPlot(immune.combined.sct, ndims = 40)
immune.combined.sct <- RunUMAP(immune.combined.sct, reduction = "pca", dims = 1:20) 
immune.combined.sct <- RunTSNE(immune.combined.sct, reduction = "pca", dims = 1:20)
immune.combined.sct <- FindNeighbors(immune.combined.sct, reduction = "pca", dims = 1:20)
immune.combined.sct <- FindClusters(immune.combined.sct, resolution = 0.6)

#Find markers
immune.combined.sct <- PrepSCTFindMarkers(immune.combined.sct)

DefaultAssay(immune.combined.sct) <- "SCT"

#Default of Wilcox
immune.combined.sct.marker.wilcox <- FindAllMarkers(immune.combined.sct, only.pos = T, min.pct = 0.25, logfc.threshold = 0.25, assay = "SCT")

immune.combined.sct.top800.wilcox<-immune.combined.sct.marker.wilcox %>% group_by(cluster) %>% top_n(n = 800, wt = avg_log2FC)

immune.combined.sct.top800.wilcox<-immune.combined.sct.top800.wilcox %>% datatable(caption="immune.combined.sct_top800_V2_Nodeletion_minc.pct0.25_logfc0.25.wilcox") %>%
  formatRound(columns=c('p_val', 'avg_log2FC','p_val_adj'),digits=3)

immune.combined.sct.top800.wilcox

**Comment by natmurad** [comment]
I had exactly the same problem with findconservedmarkers. I solved by downgrading the package.

**Comment by IPTim** [comment]
Same problem here!! Please comment!

**Comment by Qian7L** [comment]
Same problem! I noticed that there was no difference in P value between Seurat 4.2.0 and 4.1.1, the difference appeared in avg_log2FC.

**Comment by samuel-marsh** [comment]
@saketkc @timoast I can verify this issue using SeuratData and Tutorial Code.

The issue appears restricted to DE run on SCT assay.

Code here: https://gist.github.com/samuel-marsh/d117851be2ae7321e9f280e72a24d7c8
Saved marker tables as qs objects here: https://drive.google.com/drive/folders/19sVtVbnkq-XrzN2JMNN2hXilKKzX0t0o?usp=sharing

Best,
Sam

**Comment by saketkc** [comment]
Thanks for reporting this. I can confirm that the 4.2 release involved changes that were not propogated to the `FindMarkers()` function.  We will release a fix next week. Apologies for the incovenience.

**Comment by samuel-marsh** [comment]
Awesome, thanks!!!

**Comment by yamihn** [comment]
@samuel-marsh @saketkc Hi, does this problem affect the whole FindMarkers function? I'm puzzling with the fold changes calculated to search for differentially accessible peaks, whose results in terms of FoldChanges make no sense

**Comment by saketkc** [comment]
@yamihn What assay are you running `FindMarkers` on?

**Comment by lianov** [comment]
@saketkc : can you confirm that **only** version Seurat 4.2 is affected by this bug? Anything prior to it is not (e.g.: 4.1.1 etc); Want to make sure that any versions affected are flagged for us to avoid use. Thanks!

**Comment by samuel-marsh** [comment]
@lianov Appears to be and only if running using SCT assay.  Running on RNA assay following `NormalizeData` workflow was identical for me.  See GitHub gist example code above if you want to run quick tests to confirm other Seurat versions return same as 4.1.1.  It's just tutorial vignette code so it runs really quickly.

**Comment by saketkc** [comment]
Hi @lianov this bug only affects the SCT assay in Seurat v4.2. Any prior releases remain unaffected.

**Comment by ddiez** [comment]
Actually, I found this bug with the RNA assay and NormalizeData workflow (I do not use SCT). Bug present in Seurat 4.2 but problem solved if going back to 4.1.1.

**Comment by samuel-marsh** [comment]
@ddiez can you post reproducible example?

When I run `NormalizeData` pipeline from tutorial (post in code above).  I get the same results in 4.1.1 and 4.2.0

**Comment by yamihn** [comment]
> @yamihn What assay are you running `FindMarkers` on?

RNA and ATAC

**Comment by tdfy** [comment]
I can confirm different FC between 4.1.1 and 4.2.0 ( RNA ). SCTransform used to normalized/integrate; code based off of: https://hbctraining.github.io/scRNA-seq_online/schedule/links-to-lessons.html. 

100% concordance in DEGs called, FC plotted below.

![image](https://user-images.githubusercontent.com/25744474/200394679-f3a988a6-6056-4f00-810c-af309616a5db.png)

**Comment by samuel-marsh** [comment]
@tdfy yes that confirms what I posted above (SCT) assay affected.  As @saketkc mention they are working on patch.

**Comment by samuel-marsh** [comment]
@tdfy Which vignette are you following (direct link)?

**Comment by tdfy** [comment]
Running conditional differential expression per cluster/condition from RNA assay. Integration steps based on hbctraining step below. 

[https://hbctraining.github.io/scRNA-seq_online/lessons/06_integration.html](url)

```
DefaultAssay(int_seurat) <- "RNA"

clus_list <- c(0:21)

datalist =list()

for (i in clus_list){
  datalist[[paste0("lib", i)]] <- assign(paste0("lib", i),FindMarkers(int_seurat, ident.1 = "Stim",logfc.threshold = 0.09 ,verbose = TRUE, group.by="Condition", subset.ident = i)  %>% rownames_to_column(var = "gene") %>%
                                           cbind(cluster_id = i, .))
}

**Comment by samuel-marsh** [comment]
@saketkc awesome!!  thanks so much for quick fix!!

**Comment by saketkc** [comment]
Seurat v4.2.1 is now on [CRAN](https://cran.r-project.org/web/packages/Seurat/index.html). If you used `FindMarkers()` or related marker functions on the `SCT` assay, please re-run `FindMarkers` using Seurat 4.2.1+.

**Comment by nathanhaigh** [comment]
I think the fold changes are being calculated incorrectly in most places but not all. In essence, the log of the means is calculated instead of the mean of the logs. Or at least there is inconsistency in the way it is calculated.

In several places in the `FoldChange()` methods you have things like this:

```r
return(log(x = rowMeans(x = expm1(x = x)) + pseudocount.use, base = base))
```

This shows the means are calculated on the non log-transformed data and then the mean values are subsequently logged. The means should be taken on the log-transformed data. The code in `FoldChange.default()` seems to do this correctly since the fold change is calculated by subtracting the mean expression of one group from another. That would only make sense if those values were already log-transformed otherwise I'd expect a division:

```r
  # Calculate fold change
  data.1 <- mean.fxn(object[features, cells.1, drop = FALSE])
  data.2 <- mean.fxn(object[features, cells.2, drop = FALSE])
  fc <- (data.1 - data.2)
```

The consequence of calculating means before the log-transformation is that fold changes will be systematically over-estimated. For example, see: https://rpubs.com/hrlai/meanlog_logmean

**Comment by ddiez** [comment]
I still have issues with FindAllMarkers giving low fold-changes in non-SCT assay with Seurat 4.2.1. I didn't have yet time to prepare a reproducible example but I have some idea of what might be going on. Unfortunately the example shown [here](https://gist.github.com/samuel-marsh/d117851be2ae7321e9f280e72a24d7c8) is not comparing apples with apples. While the SCT pipeline is doing integration, the standard pipeline isn't. In my case the issue is in a dataset for which I have done integration (not Seurat integration but with another tool). So I have an "integrated" assay. Doing FindAllMarkers on RNA assay with Seurat 4.2.x gives lower FC than using 4.1.1. I think having an "integrated" assay might be the root of the issue.

I will nonetheless try to put a reproducible example later. It should be simple to modify the code above to include Seurat integration in the standard pipeline.

**Comment by samuel-marsh** [comment]
HI @ddiez,

If you can provide reproducible example that would be great.  I just ran new test and confirmed no difference between 4.1.1 and 4.2.1.  Running SCT and Integration then switching to RNA assay and normalizing data before running marker analysis.  New code with results pasted in is here:
https://gist.github.com/samuel-marsh/d32b79421773587bce8be1feb78a5ab2

Best,
Sam

**Comment by ddiez** [comment]
Hi @samuel-marsh,

Thanks but in that gist the scenario I described is not included. Basically I do not use SCT and perhaps in the case of SCT assays/integration the issue has been solved. My point is that this still happens when other integrations are present and this has not been solved. The following code reproduces the issue: 

https://gist.github.com/ddiez/9ebae2c95a9bcb57554c4fc0c28460c2

Basically I run everything with 4.2.1 up to FindMarkers, then switch to 4.1.1 and run FindMarkers again. Even though nothing else has changed in the Seurat object the fold changes are significantly smaller when FindMakers from 4.2.1 is used.

Let me know if this isn't clear or want me to modify the code. Thanks!

Best,
Diego

**Comment by ddiez** [comment]
Thanks @samuel-marsh for the additional comparison. I was puzzled by the fact that you had to run `NormalizeData` before `FindMarkers` to recover the behavior in 4.1.1. I checked the value in `ifnb@assays$RNA@data` before and after running NormalizeData and as expected they were identical.

Then I checked the value for `Command(ifnb, "NormalizeData", "normalization.method")` before integration and after integration, and I got the following results:

Before integration:

```
> lapply(ifnb, Command, command="NormalizeData.RNA", value="normalization.method")
$CTRL
[1] "LogNormalize"

$STIM
[1] "LogNormalize"
```

After integration:

```
> Command(ifnb, "NormalizeData.RNA", value="normalization.method")
Error in Command.Seurat(ifnb, "NormalizeData.RNA", value = "normalization.method") : 
  NormalizeData.RNA has not been run or is not a valid command.
```

So the normalization method log is not propagated to the integrated object. This is with Seurat 4.1.1 so probably is what always has happened. But It seem that this command is used to identify the normalization.method that will be passed to `FoldChange.Assay()`. Probably before 4.2.0 it didn't matter that the normalization method command was not propagated to the integrated object but I noticed that changes on how this is handled were introduced in commit 71193c0fecab562f8495f6c0eb2873e5f75dccce. So maybe there is the issue?

**Comment by samuel-marsh** [comment]
@saketkc and Seurat team,

Reopening this issue here so additional issue described in previous 2 posts can be addressed.

Thanks,
Sam

**Comment by saketkc** [comment]
Thanks  @ddiez and @samuel-marsh for providing a reproducible example. I can confirm the issue. We will have a fix soon.

**Comment by samuel-marsh** [comment]
Thank you @saketkc !! :)

**Comment by saketkc** [comment]
We have released a fix for the issue mentioned in  https://github.com/satijalab/seurat/issues/6586#issuecomment-1311309196 in the develop branch through https://github.com/satijalab/seurat/commit/0e94e18f6bafb9737305ad9779d42cbb2487c040 which is also part of Seurat 4.3 (already on CRAN). Thanks for your patience.

**Comment by samuel-marsh** [comment]
@saketkc fantastic!! Thanks so much! 😄
